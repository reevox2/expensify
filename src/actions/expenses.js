import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

//ASYNC DATABASE CREATE + send to redux reducer
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = (expense) => {
  return dispatch => {
    return database.ref(`/expenses/${expense.id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense(expense))
      })
      .catch(e => console.log(e));
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

//SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    // populate expenses
    return database.ref('expenses').once('value').then(snapshot => {
      const expenses = [];
      snapshot.forEach(expenseSnapshot => {
        expenses.push({
          id: expenseSnapshot.key,
          ...expenseSnapshot.val()
        });
      })
      dispatch(setExpenses(expenses));
    })
  }
};
