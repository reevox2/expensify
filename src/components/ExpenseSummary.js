import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpenseSummary = (props) => (
    <div>
        <p>
            Viewing {props.expenses.length} expense{props.expenses.length === 1 ? '' : 's'} totaling {numeral(props.expensesTotal).format('$0,0.00')}.
        </p>
    </div>
);


const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters),
    expensesTotal: selectExpensesTotal(state.expenses)/100
})

export default connect(mapStateToProps)(ExpenseSummary);