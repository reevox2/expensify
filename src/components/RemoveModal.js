import Modal from 'react-modal';
import React from 'react';

Modal.setAppElement('#app');

export default (props) => (
    <Modal
        isOpen={props.modalOpen}
        contentLabel="Remove Expense"
        closeTimeoutMS={200}
        className="modal"
        onRequestClose={props.closeModal}
    >
        <h1 className="modal__title">Are you sure you?</h1>
        <button onClick={props.remove} className="modal__body button">Yup. Remove.</button>
    </Modal>
);