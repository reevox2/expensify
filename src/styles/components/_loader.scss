.loader {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    width: 100wh;
}

.loader__image {
    height: 6rem;
    width: 6rem;
}