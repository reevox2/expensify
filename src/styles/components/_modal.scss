.ReactModalPortal>div {
    opacity: 0;
}
  
.ReactModalPortal .ReactModal__Overlay {
    align-items: center;
    display: flex;
    justify-content: center;
    transition: opacity 200ms ease-in-out;
}

.ReactModalPortal .ReactModal__Overlay--after-open {
    opacity: 1;
}

.ReactModalPortal .ReactModal__Overlay--before-close {
    opacity: 0;
}

.modal {
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: $dark-blue;
    color: $off-white;
    justify-content: center;
    padding: $m-size;
}

.modal__title {
    margin: 0 0 $s-size 0;
}

.modal__body {
    margin: 0;
}