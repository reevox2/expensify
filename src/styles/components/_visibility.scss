.show-for-mobile {
  @media (min-width: $desktop-breakpoint) {
    display: none;
  }
}

.show-for-mobile {
  @media (max-width: $desktop-breakpoint - 0.01rem) {
    display: none;
  }
}
