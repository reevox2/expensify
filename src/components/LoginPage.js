import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startFacebookLogin, startLink } from '../actions/auth';

export const LoginPage = ({ startLogin, startFacebookLogin, performLink, startLink }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            {(performLink) ? (
                <div>
                    <p>You must link to your old account, to be able to use both sign in methods.</p>
                    <button className="button" onClick={startLink}>Link your Accounts</button>
                </div>
            ) : (
                    <div>
                        <p>It's time to get your expenses under control.</p>
                        <button className="button" onClick={startLogin}>Login with Google</button>
                        <button className="button button--facebook" onClick={startFacebookLogin}>Login with Facebook</button>
                    </div>
                )}

        </div>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
    startFacebookLogin: () => dispatch(startFacebookLogin()),
    startLink: () => dispatch(startLink()),
});

const mapStateToProps = (state) => ({
    performLink: (state.auth.user !== undefined)
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);