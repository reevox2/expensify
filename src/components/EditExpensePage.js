import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import RemoveModal from './RemoveModal';

export class EditExpensePage extends React.Component {
  state = {
    modalOpen: false
  };
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  confirmRemove = () => {
    console.log(this.state.modalOpen);
    this.setState(state => ({ modalOpen: !state.modalOpen }));
  }
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container list-body">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.confirmRemove}>
            Remove Expense
          </button>
          <RemoveModal modalOpen={this.state.modalOpen} remove={this.onRemove} closeModal={this.confirmRemove} />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
