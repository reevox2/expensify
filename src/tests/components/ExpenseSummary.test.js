import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';



test('should render correctly with only one expense', ()=>{
    const wrapper = shallow(<ExpenseSummary expenses={[expenses[0]]} expensesTotal={94.34} />);
    expect(wrapper).toMatchSnapshot();
})

test('should render correctly with multiple expenses', ()=>{
    const wrapper = shallow(<ExpenseSummary expenses={expenses} expensesTotal={94.34} />);
    expect(wrapper).toMatchSnapshot();
})