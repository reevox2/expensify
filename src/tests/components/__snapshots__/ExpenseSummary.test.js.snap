// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totaling 
      <span>
        $23,141,231,241.23
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render correctly with only one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totaling 
      <span>
        $94.34
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
